File: app/src/main/java/fr/nihilus/mymusic/view/CurrentlyPlayingDecoration.java
package fr.nihilus.mymusic.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.drawable.Drawable;
import android.support.annotation.ColorInt;
import android.support.annotation.IntRange;
import android.support.annotation.NonNull;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.support.v7.content.res.AppCompatResources;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import fr.nihilus.mymusic.R;

public class CurrentlyPlayingDecoration extends RecyclerView.ItemDecoration {

    private final int mPaddingStart;
    private Drawable mIcon;
    private int mDecoratedPosition;

    public CurrentlyPlayingDecoration(@NonNull Context context) {
        mIcon = AppCompatResources.getDrawable(context, R.drawable.currently_playing_decoration);
        mDecoratedPosition = RecyclerView.NO_POSITION;
        mPaddingStart = context.getResources().getDimensionPixelSize(R.dimen.list_item_horizontal_padding);
    }

    @Override
    public void onDrawOver(Canvas c, RecyclerView parent, RecyclerView.State state) {
        if (mDecoratedPosition != RecyclerView.NO_POSITION) {
            final View child = parent.getChildAt(mDecoratedPosition);
            if (child != null) {
                int left = mPaddingStart;
                int top = child.getTop() + (child.getHeight() - mIcon.getIntrinsicHeight()) / 2;
                int right = left + mIcon.getIntrinsicWidth();
                int bottom = top + mIcon.getIntrinsicHeight();

                mIcon.setBounds(left, top, right, bottom);
                mIcon.draw(c);
            }
        }
    }

    public void setDecoratedItemPosition(@IntRange(from = -1) int position) {
        mDecoratedPosition = position;
    }

    public void setIconColor(@ColorInt int color) {
        mIcon = DrawableCompat.wrap(mIcon);
        DrawableCompat.setTint(mIcon, color);
    }
}

File: app/src/main/java/fr/nihilus/music/view/CurrentlyPlayingDecoration.kt
/*
 * Copyright 2018 Thibault Seisel
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.nihilus.music.view

import android.content.Context
import android.graphics.Canvas
import android.graphics.drawable.Drawable
import android.support.v7.widget.RecyclerView
import fr.nihilus.music.R

class CurrentlyPlayingDecoration(
    context: Context,
    iconColor: Int
) : RecyclerView.ItemDecoration() {

    private val icon: Drawable = context.getDrawable(R.drawable.currently_playing_decoration)
            ?: throw IllegalStateException("Icon should not be null")
    private val paddingStart =
        context.resources.getDimensionPixelSize(R.dimen.list_item_horizontal_padding)
    var decoratedPosition = RecyclerView.NO_POSITION

    init {
        icon.setTint(iconColor)
    }

    override fun onDrawOver(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        if (decoratedPosition != RecyclerView.NO_POSITION) {
            val child = parent.getChildAt(decoratedPosition)
            if (child != null) {
                val left = paddingStart
                val top = child.top + (child.height - icon.intrinsicHeight) / 2
                val right = left + icon.intrinsicWidth
                val bottom = top + icon.intrinsicHeight

                icon.setBounds(left, top, right, bottom)
                icon.draw(c)
            }
        }
    }
}
